src/transformers/trainer_tf.py

```diff
@@ -0,0 +1,509 @@
+            logger.info("Loading features from cached file %s", cached_validation_features_file)
+            self.datasets["validation"] = self._load_cache(cached_validation_features_file)
+            logger.info("Loading features from cached file %s", cached_test_features_file)
+            self.datasets["test"] = self._load_cache(cached_test_features_file)
+        else:
+            os.makedirs(cache_dir, exist_ok=True)
+            self._create_features(tokenizer)
+            logger.info("Create cache file %s", cached_training_features_file)
+            self._save_cache("train", cached_training_features_file)
+            logger.info("Create cache file %s", cached_validation_features_file)
+            self._save_cache("validation", cached_validation_features_file)
+            logger.info("Create cache file %s", cached_test_features_file)
+            self._save_cache("test", cached_test_features_file)
+
+        self.train_steps = math.ceil(len(list(self.datasets["train"].as_numpy_iterator())) / self.batch_size)
+        self.datasets["train"] = self.datasets["train"].shuffle(128).batch(self.batch_size).repeat(-1)
+        self.datasets["train"] = self.strategy.experimental_distribute_dataset(self.datasets["train"])
+        self.validation_steps = math.ceil(len(list(self.datasets["validation"].as_numpy_iterator())) / self.eval_batch_size)
+        self.datasets["validation"] = self.datasets["validation"].batch(self.eval_batch_size)
+        self.datasets["validation"] = self.strategy.experimental_distribute_dataset(self.datasets["validation"])
+        self.test_steps = math.ceil(len(list(self.datasets["test"].as_numpy_iterator())) / (self.eval_batch_size / self.strategy.num_replicas_in_sync))
+        self.datasets["test"] = self.datasets["test"].batch(self.eval_batch_size // self.strategy.num_replicas_in_sync)
+
+    def _create_optimizer(self):
+        """
+        Create the training optimizer with its name. Allowed names:
+          - adam: Adam optimizer
+          - adamw: Adam with Weight decay optimizer
+          - adadelta: Adadelta optimizer
+          - rmsprop: Root Mean Square Propogation optimizer
+          - sgd: Stochastic Gradient Descent optimizer
+        """
+        if self.optimizer_name == "adamw":
+            if self.decay_steps is None:
+                self.decay_steps = self.train_steps
+
+            learning_rate_fn = tf.keras.optimizers.schedules.PolynomialDecay(initial_learning_rate=self.learning_rate,
+                                                                             decay_steps=self.decay_steps,
+                                                                             end_learning_rate=0.0)
+            if self.warmup_steps:
+                learning_rate_fn = WarmUp(initial_learning_rate=self.learning_rate, decay_schedule_fn=learning_rate_fn,
+                                          warmup_steps=self.warmup_steps)
```

Comment:
Should we allow this for all optimizers?

Maybe we could even let the user supply his own optimizer? Maybe directly an instance or maybe supplying an identifier for [`tf.keras.optimizers.get()`](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/get)?

Usually in the library we allow for two paths:
- one very simple path with good defaults (here it could be `adamw` with good schedule).
- one fully adaptable path where you can do what you want (here supply your own optimizer).

Reply:
It is exaclty what I have in mind, (the same goes for the loss as well) but in order to focus on the trainer logic mainly I decided to go for the easiest way of implementation. But, definitely the direction to take is what you haved described.